### DISCUSSION TITLE
Replace _id to id

### DISCUSSION THREAD

**Discussion by edw19** [typegoose/typegoose#596] [Q&A] [answered]
I am testing typegoose and I need to know if it is possible to do this conversion
I try this solution from stack overflow but dont working

https://stackoverflow.com/questions/60804946/replace-id-to-id-in-nestjs-typegoose?newreg=13f9f18d9e53422e8b7d3fff677cdba4

**Comment by smolinari** [comment]
Those SO answers are old. 

You can implement the `Base` interface and add the `id` and  `_id` properties (without the `@Prop` decorator). 

```ts
import { Base } from '@typegoose/typegoose/lib/defaultClasses';

export class MyEntity implements Base {
  _id!: Types.ObjectId;

  id!: string;

  @Prop({ required: true })
  title!: string;

  @Prop()
  description?: string;

  @Prop({ required: true })
  completed!: boolean;
  
  ...
  ```
  
 See if that gets you what you need. 
  
 Scott

**Reply by edw19** [reply]
dont work, this is my model 

![image](https://user-images.githubusercontent.com/47750859/131013032-84037912-2093-4b86-8e10-a59f50e01e8f.png)

and result its the same

![image](https://user-images.githubusercontent.com/47750859/131013960-ad68b09f-3797-4254-93f3-2439338a84c4.png)

**Comment by smolinari** [comment, marked as answer]
@edw19 - What am I looking at as the result above? The `id` field is a virtual. It won't be in the database. It's a virtual of `_id`. 

https://mongoosejs.com/docs/guide.html#id

Scott

**Reply by hasezoey** [reply]
also a addition, from what i know, `_id` is always needed in mongodb / mongoose, so even if you (@edw19) want to change `id` to not be a virtual and have it act as an actual property, then `_id` would still be there

[also, the `_id` type can be changed](https://typegoose.github.io/typegoose/docs/guides/advanced/change-id-type)

**Reply by edw19** [reply]
I think I got the question wrong, what I want to get is that the _id property is read by graphql as id, so the following example shows my case
![image](https://user-images.githubusercontent.com/47750859/131158419-ed7ce395-d70a-41d8-b105-3b5239f2a3d1.png)

but in graphql it doesn't recognize the variable .id because it shows _id

![image](https://user-images.githubusercontent.com/47750859/131158595-db23f01a-2667-484f-ab11-145339ef7303.png)

![image](https://user-images.githubusercontent.com/47750859/131158714-92d8e7fc-2388-4178-a49c-cb70853b6f84.png)

![image](https://user-images.githubusercontent.com/47750859/131158750-3093939f-b5f6-4996-af7e-3100eb9de515.png)

**Reply by hasezoey** [reply]
@edw19 if you want the normal `id` virtual to show up in the output (like `toObject` or `toJSON`) you have to enable it, either when converting or as schemaOptions

see [Why is my virtual not included in the output?](https://typegoose.github.io/typegoose/docs/api/virtuals/#why-is-my-virtual-not-included-in-the-output)

**Reply by edw19** [reply]
thanks that was what i was looking for 😊😊😊😊👌

**Reply by smolinari** [reply]
Darnit. I came close to writing that too, but I thought we were talking about the entity itself, not the output object. Thanks @hasezoey. 

Scott

**Reply by igortas** [reply]
Still not working

**Reply by smolinari** [reply]
@igortas - You should start a new issue and refer back to this one.

Scott
